File: food-ordering-system/restaurant-service/src/main/java/mk/finki/restaurantservice/domain/Restaurant.java
package mk.finki.restaurantservice.domain;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "restaurants")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
public class Restaurant {
    @Id
    private UUID id;

    private String name;
    private String address;
    private boolean active;

    @OneToMany(mappedBy = "restaurant", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<MenuItem> menu = new ArrayList<>();

    public void addMenuItem(String name, BigDecimal price) {
        // Enforce business rule: No duplicate names within a restaurant's menu
        boolean exists = menu.stream().anyMatch(item -> item.getName().equalsIgnoreCase(name));
        if (exists) {
            throw new IllegalArgumentException("Menu item with name " + name + " already exists");
        }
        MenuItem newItem = new MenuItem(name, price, this);
        this.menu.add(newItem);
    }
}

File: food-ordering-system/restaurant-service/src/main/kotlin/mk/finki/restaurantservice/domain/restaurant/Restaurant.kt
package mk.finki.restaurantservice.domain.restaurant

import io.swagger.v3.oas.annotations.media.Schema
import jakarta.persistence.*
import mk.finki.restaurantservice.domain.restaurant.entities.MenuItem
import mk.finki.restaurantservice.domain.restaurant.enums.MenuItemStatus
import mk.finki.restaurantservice.domain.restaurant.exceptions.DuplicateMenuItemException
import mk.finki.restaurantservice.domain.restaurant.valueobjects.MenuItemId
import mk.finki.restaurantservice.domain.restaurant.valueobjects.Money
import mk.finki.restaurantservice.domain.restaurant.valueobjects.RestaurantId
import java.util.UUID

@Entity
@Table(name = "restaurants")
class Restaurant(
    @EmbeddedId
    @Schema(example = "123e4567-e89b-12d3-a456-426614174000")
    val id: RestaurantId,
    @Schema(example = "Pizza Palace")
    var name: String,
    @Schema(example = "456 Oak St, London")
    var address: String
) {
    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    @JoinColumn(name = "restaurant_id")
    private val _menu: MutableList<MenuItem> = mutableListOf()

    @Schema(example = "[]")
    val menu: List<MenuItem>
        get() = _menu.filter { !it.deleted }

    fun addMenuItem(name: String, price: Money): MenuItem {
        if (_menu.any { it.name == name && !it.deleted }) {
            throw DuplicateMenuItemException("Menu item with name $name already exists in restaurant ${this.name}")
        }
        val menuItem = MenuItem(
            id = MenuItemId(UUID.randomUUID()),
            name = name,
            price = price,
            status = MenuItemStatus.AVAILABLE
        )
        _menu.add(menuItem)
        return menuItem
    }

    fun discontinueMenuItem(menuItemId: MenuItemId) {
        _menu.find { it.id == menuItemId }?.discontinue()
    }
    
    // For JPA
    protected constructor() : this(RestaurantId(UUID.randomUUID()), "", "")
}
